spec/base_spec.rb

```diff
@@ -0,0 +1,97 @@
+# frozen_string_literal: true
+
+RSpec.describe IOPromise::Base do
+  # Implement the very minimal requirements of an IOPromise::Base subclass
+  class DummyExecutorPool < IOPromise::ExecutorPool::Sequential
+  end
+  class DummyPromise < IOPromise::Base
+    def execute_pool
+      DummyExecutorPool.for(Thread.current)
+    end
+  end
+
+  it "fulfills like a normal promise" do
+    b = DummyPromise.new
+    chained = b.then { |v| v }
+    expect(b).to be_pending
+    expect(chained).to be_pending
+
+    b.fulfill('test')
+
+    expect(b).to_not be_pending
+    expect(b).to be_fulfilled
+    expect(b.value).to eq('test')
+
+    expect(chained).to_not be_pending
+    expect(chained).to be_fulfilled
+    expect(chained.value).to eq('test')
+  end
+
+  it "rejects like a normal promise" do
+    b = DummyPromise.new
+    chained = b.then { |v| v }
+    expect(b).to be_pending
+    expect(chained).to be_pending
+
+    b.reject('test')
+
+    expect(b).to_not be_pending
+    expect(b).to be_rejected
+    expect(b.reason).to eq('test')
+
+    expect(chained).to_not be_pending
+    expect(chained).to be_rejected
+    expect(chained.reason).to eq('test')
+  end
+
+  context "#cancel" do
+    it "prevents handlers on fulfill" do
+      b = DummyPromise.new
+      chained = b.then { |v| v }
+      expect(b).to be_pending
+      expect(chained).to be_pending
+      expect(b).to_not be_cancelled
+
+      b.cancel
+
+      expect(b).to be_cancelled
+
+      b.fulfill('test')
+
+      expect(b).to be_pending
+      expect(chained).to be_pending
+    end
+
+    it "prevents handlers on reject" do
+      b = DummyPromise.new
+      chained = b.then { |v| v }
+      expect(b).to be_pending
+      expect(chained).to be_pending
+      expect(b).to_not be_cancelled
+
+      b.cancel
+
+      expect(b).to be_cancelled
+
+      b.reject('test')
+
+      expect(b).to be_pending
+      expect(chained).to be_pending
+    end
+
+    it "does nothing on completed promises" do
+      b = DummyPromise.new
+      b.fulfill('foo')
+
+      expect(b).to_not be_pending
+      expect(b).to be_fulfilled
+      expect(b).to_not be_cancelled
+
+      b.cancel
+
+      expect(b).to_not be_pending
```

Comment:
What happens if I do something like this after cancelling `b`?

```ruby
b.then { |v| v }
```

Does the chained promise execute? Should we override `then` as well? I might be getting a bit too paranoid with this suggestion 😄

Reply:
In the current implementation, cancelling does nothing once a promise is resolved, which I think is acceptable since it's not a time delayed execution from the shared IOPromise event loop at that point. I believe the only time there's a risk is where a promise ends up pending but not synced, inside the IOPromise event loop.

Reply:
I'm going to add a test for `.then` in the case where cancellation happens (before resolving) and the case where cancellation doesn't do anything (after resolving), which makes this behaviour explicit in the form of a test. I feel like changing `cancel` to act on already-resolved promises could get a bit more complex (since actions up to that point already completed) but if there's a case where this would prevent an issue it might make sense?